### PR TITLE
[DT-294-524-525] tanagra aou rt/ct zipcode-genomics-visits-device underlay updates 

### PR DESCRIPTION
**Pull request by chenchals**
This PR addresses the following tickets for indexing build Underlays that define CDR:
- [DT-294](https://precisionmedicineinitiative.atlassian.net/browse/DT-294) Visit and Device domains
- [DT-524](https://precisionmedicineinitiative.atlassian.net/browse/DT-524) Genomics (CT only)
- [DT-525](https://precisionmedicineinitiative.atlassian.net/browse/DT-525) Zipcode Socioeconomic PrePackaged (CT only)
- Underlays and ui/top-level.json were updated for
  - CT (SC2023Q3R1) to include
    - Genomics : Cohort criteria
    - Zipcode Socioeconomic : PrePackaged Concept Set
    - Visit : Cohort criteria, Concept Set
    - Device : Cohort criteria, Concept Set
  - RT (SR2023Q3R1) to include
    - Visit : Cohort criteria, Concept Set
    - Device : Cohort criteria, Concept Set
    
After merging the PR, I will publish the datasets to secure perimeter projects for RT and CT along with updating the tanagra module in AoU

### CODE REVIEW

File: service/src/main/resources/config/aou/test/SC2023Q3R1/entity/t_device.json [hunk 1]
```diff
@@ -4,42 +4,18 @@
   "attributes": [
     { "type": "SIMPLE", "name": "id", "dataType": "INT64" },
     { "type": "SIMPLE", "name": "name", "dataType": "STRING" },
-    { "type": "KEY_AND_DISPLAY", "name": "vocabulary", "dataType": "STRING" },
-    { "type": "KEY_AND_DISPLAY", "name": "standard_concept", "dataType": "STRING" },
+    { "type": "SIMPLE", "name": "vocabulary_id", "dataType": "STRING" },
+    { "type": "SIMPLE", "name": "vocabulary", "dataType": "STRING" },
```
Comment (marikomedlock) [inline service/src/main/resources/config/aou/test/SC2023Q3R1/entity/t_device.json:8]: Why the change from a single `KEY_AND_DISPLAY` vocabulary attribute to two separate `SIMPLE` attributes?

File: service/src/main/resources/config/aou/test/SC2023Q3R1/entitygroup/t_device_person_occurrence.json [hunk 2]
```diff
@@ -20,6 +20,11 @@
         "idPairsTable": { "table": "device_exposure" },
         "idPairsIdA": { "column": "device_concept_id" },
         "idPairsIdB":  { "column": "person_id" }
+      },
+      "groupToItems": {
```
Comment (marikomedlock) [inline service/src/main/resources/config/aou/test/SC2023Q3R1/entitygroup/t_device_person_occurrence.json:24]: This property doesn't make sense here. The `type` above says `CRITERIA_OCCURRENCE`, not `GROUP_ITEMS`. I think you just want to leave this as is, but what were you trying to add here?

File: service/src/main/resources/config/aou/test/SC2023Q3R1/entitygroup/t_visit_person_occurrence.json [hunk 3]
```diff
@@ -20,6 +20,11 @@
         "idPairsTable": { "table": "visit_occurrence" },
         "idPairsIdA": { "column": "visit_concept_id" },
         "idPairsIdB":  { "column": "person_id" }
+      },
+      "groupToItems": {
```
Comment (marikomedlock) [inline service/src/main/resources/config/aou/test/SC2023Q3R1/entitygroup/t_visit_person_occurrence.json:24]: Ditto above for device.

File: service/src/main/resources/config/aou/test/SC2023Q3R1/entitygroup/t_zipcode_socioeconomic_person.json [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  "type": "GROUP_ITEMS",
+  "groupEntity": "t_person",
+  "itemsEntity": "t_zipcode_socioeconomic",
+  "sourceDataMapping": {
+    "dataPointer": "omop_dataset",
+    "relationshipMappings": {
+      "groupToItems": {
+        "idPairsTable": { "rawSqlFile": "t_zipcode_socioeconomic.sql" },
+        "idPairsIdA": { "column": "person_id" },
+        "idPairsIdB":  { "column": "person_id" }
```
Comment (marikomedlock) [inline service/src/main/resources/config/aou/test/SC2023Q3R1/entitygroup/t_zipcode_socioeconomic_person.json:12]: This doesn't look right. The entity group definitions are basically defining JOINs between tables, so it doesn't make sense to join the same two fields on each other.

File: service/src/main/resources/config/aou/test/SC2023Q3R1/sql/t_device.sql [hunk 5]
```diff
@@ -0,0 +1,19 @@
+SELECT DISTINCT c.concept_id, c.concept_name, c.concept_code,
+                CASE
+                    WHEN c.standard_concept IS NULL THEN 'Source'
+                    WHEN c.standard_concept = 'S' THEN 'Standard'
+                    ELSE 'Unknown' END standard_concept,
+                v.vocabulary_name,
+                v.vocabulary_id
+
+FROM `all-of-us-ehr-dev.SC2023Q3R1.device_exposure` AS de
```
Comment (marikomedlock) [inline service/src/main/resources/config/aou/test/SC2023Q3R1/sql/t_device.sql:9]: Here you're selecting all the device types that have >=1 occurrence row, instead of selecting just all device types. Is that what you intended? Or what was the motivation behind the changes to the device entity?

File: service/src/main/resources/config/aou/test/SC2023Q3R1/sql/t_visit.sql [hunk 6]
```diff
@@ -0,0 +1,11 @@
+SELECT
+  distinct c.concept_id, c.concept_name
+
+FROM `all-of-us-ehr-dev.SC2023Q3R1.visit_occurrence` AS vo
```
Comment (marikomedlock) [inline service/src/main/resources/config/aou/test/SC2023Q3R1/sql/t_visit.sql:4]: Ditto comment for device sql above.

File: service/src/main/resources/config/aou/test/SC2023Q3R1/entity/t_zipcode_socioeconomic.json [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    { "type": "SIMPLE", "name": "vacant_housing", "dataType": "DOUBLE" },
+    { "type": "SIMPLE", "name": "deprivation_index", "dataType": "DOUBLE" },
+    { "type": "SIMPLE", "name": "american_community_survey_year", "dataType": "INT64" }
+  ],
+
+  "sourceDataMapping": {
+    "dataPointer": "omop_dataset",
+    "tablePointer": { "rawSqlFile": "t_zipcode_socioeconomic.sql" },
+    "attributeMappings": {
+      "id": { "value": { "column": "person_id" } }
```
Comment (marikomedlock) [inline service/src/main/resources/config/aou/test/SC2023Q3R1/entity/t_zipcode_socioeconomic.json:23]: I would expect this to be the unique id/primary key of the related table, not the `person` table. How many rows in `zip3_ses_map` and/or `observation` filtered on the zip concept id/strings are you expecting per `person`?

File: service/src/main/resources/config/aou/test/SC2023Q3R1/entitygroup/t_device_person_occurrence.json (hunk 2 above)
Comment (chenchals) [reply service/src/main/resources/config/aou/test/SC2023Q3R1/entitygroup/t_device_person_occurrence.json:24]: Not sure... I do not understand this node... but hey it is working. Can you fix this? I do not understand the relationshipMapping. Need more documenttation

File: service/src/main/resources/config/aou/test/SC2023Q3R1/entitygroup/t_zipcode_socioeconomic_person.json (hunk 4 above)
Comment (chenchals) [reply service/src/main/resources/config/aou/test/SC2023Q3R1/entitygroup/t_zipcode_socioeconomic_person.json:12]: Not sure what is correct. I just need the [zipcode sql ](https://github.com/DataBiosphere/tanagra/blob/19af2091a9f745db3685d10c12356407b94d6a81/service/src/main/resources/config/aou/test/SC2023Q3R1/sql/t_zipcode_socioeconomic.sql#L1-L1) table to be a prepackaged concept setup.

File: service/src/main/resources/config/aou/test/SC2023Q3R1/sql/t_device.sql (hunk 5 above)
Comment (chenchals) [reply service/src/main/resources/config/aou/test/SC2023Q3R1/sql/t_device.sql:9]: In Aou we do not show *all* whe show only those device or any other domain concept_ids only if there is atleast 1 obeservation or item_count or rollup count is >=1 This part is from aou's indexing build SQL: [build-cb-criteria-device.sh](https://github.com/all-of-us/workbench/blob/a909b2097bfa407c050fdaa5b2b351a212f4f8c1/api/db-cdr/generate-cdr/build-cb-criteria-device.sh#L67-L71)

File: service/src/main/resources/config/aou/test/SC2023Q3R1/sql/t_visit.sql (hunk 6 above)
Comment (chenchals) [reply service/src/main/resources/config/aou/test/SC2023Q3R1/sql/t_visit.sql:4]: Same, if a particular visit is not observed (no-person has such a visit) we do not show that visit ref:[build-cb-criteria-visit.sh](https://github.com/all-of-us/workbench/blob/a909b2097bfa407c050fdaa5b2b351a212f4f8c1/api/db-cdr/generate-cdr/build-cb-criteria-visit.sh#L57-L63). Any row where concept_id counts is 0 are removed see [cb_criteria-clean-up](https://github.com/all-of-us/workbench/blob/a909b2097bfa407c050fdaa5b2b351a212f4f8c1/api/db-cdr/generate-cdr/build-cb-criteria-attribute-tables-and-cleanup.sh#L1-L1)

File: service/src/main/resources/config/aou/test/SC2023Q3R1/entity/t_zipcode_socioeconomic.json (hunk 7 above)
Comment (chenchals) [reply service/src/main/resources/config/aou/test/SC2023Q3R1/entity/t_zipcode_socioeconomic.json]: Updated tables and SQL for t_zipcopde_socioeconomic

File: service/src/main/resources/config/aou/test/SC2023Q3R1/entitygroup/t_zipcode_socioeconomic_person.json (hunk 4 above)
Comment (chenchals) [reply service/src/main/resources/config/aou/test/SC2023Q3R1/entitygroup/t_zipcode_socioeconomic_person.json]: Updated t_zipcode_socionomic tables and SQL

File: service/src/main/resources/config/aou/test/SC2023Q3R1/entitygroup/t_device_person_occurrence.json (hunk 2 above)
Comment (chenchals) [reply service/src/main/resources/config/aou/test/SC2023Q3R1/entitygroup/t_device_person_occurrence.json]: Git it. removed `groupToItems` node

File: service/src/main/resources/config/aou/test/SC2023Q3R1/entity/t_device.json (hunk 1 above)
Comment (chenchals) [reply service/src/main/resources/config/aou/test/SC2023Q3R1/entity/t_device.json:8]: btw, also in AoU, dataset outputs we provide both `concept_id` and `concept_name` for all concepts - gender_concept_id etc.. All these are fixes for the second round, once we get the cohort building to work as exoected in tanagra for AoU dataset.

File: service/src/main/resources/config/aou/test/SC2023Q3R1/sql/t_zipcode_socioeconomic.sql [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    zip_code.median_income,
+    zip_code.fraction_no_health_ins as no_health_insurance,
+    zip_code.fraction_poverty as poverty,
+    zip_code.fraction_vacant_housing as vacant_housing,
+    zip_code.deprivation_index,
+    zip_code.acs as american_community_survey_year
+FROM
+    `all-of-us-ehr-dev.SC2023Q3R1.zip3_ses_map` zip_code
+JOIN
+    `all-of-us-ehr-dev.SC2023Q3R1..observation` observation
```
Comment (marikomedlock) [inline service/src/main/resources/config/aou/test/SC2023Q3R1/sql/t_zipcode_socioeconomic.sql:18]: Two periods?

**Review by marikomedlock — APPROVED** [review]

File: service/src/main/resources/config/aou/test/SC2023Q3R1/sql/t_zipcode_socioeconomic.sql (hunk 8 above)
Comment (chenchals) [reply service/src/main/resources/config/aou/test/SC2023Q3R1/sql/t_zipcode_socioeconomic.sql]: good catch fixed it.
